modules/fundamental/src/analysis/service/mod.rs

```diff
@@ -0,0 +1,584 @@
+use crate::Error;
+use sea_orm::{prelude::ConnectionTrait, Statement};
+use std::collections::HashMap;
+use tracing::instrument;
+use trustify_common::{
+    db::{query::Query, Database, Transactional},
+    model::{Paginated, PaginatedResults},
+};
+
+use petgraph::graph::{Graph, NodeIndex};
+// use petgraph::visit::EdgeRef;
+use petgraph::Direction;
+use regex::Regex;
+use serde::Serialize;
+use std::fmt;
+use std::str::FromStr;
+use trustify_common::purl::Purl;
+use trustify_entity::relationship::Relationship;
+use utoipa::ToSchema;
+
+#[derive(Debug, Clone, PartialEq, Eq, ToSchema, serde::Serialize)]
+pub struct PackageNode {
+    purl: String,
+    name: String,
+    published: String,
+    document_id: String,
+    product_name: String,
+    product_version: String,
+}
+impl fmt::Display for PackageNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.name)
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct AdvisoryGraphSummary {
+    pub component: String,
+    pub ancestors: Vec<PackageNode>,
+}
+impl AdvisoryGraphSummary {
+    fn new(component: String, ancestors: Vec<PackageNode>) -> AdvisoryGraphSummary {
+        AdvisoryGraphSummary {
+            component,
+            ancestors,
+        }
+    }
+}
+
+pub struct AnalysisService {
+    db: Database,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn add_node(
+    g: &mut Graph<PackageNode, Relationship, petgraph::Directed>,
+    nodes: &mut HashMap<String, NodeIndex>,
+    component_purl: &String,
```

Comment:
suggestion: `&str` instead of `&String`.  Calling it with a reference to a `String` satisfies `&str` as does `"string literals"` generally used in tests.

Only ask for non-ref `String` if you want to take ownership, else `&str`.